runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/StreamingEngineClient.java

```diff
@@ -0,0 +1,464 @@
+    if (getWorkerMetadataReady.getCount() > 0) {
+      getWorkerMetadataReady.countDown();
+    } else {
+      requestBudgetRefresh();
+    }
+  }
+
+  public ImmutableList<Long> getAndResetThrottleTimes() {
+    StreamEngineConnectionState currentConnections = connections.get();
+
+    ImmutableList<Long> keyedWorkStreamThrottleTimes =
+        currentConnections.windmillStreams().values().stream()
+            .map(WindmillStreamSender::getAndResetThrottleTime)
+            .collect(toImmutableList());
+
+    return ImmutableList.<Long>builder()
+        .add(getWorkerMetadataThrottleTimer.getAndResetThrottleTime())
+        .addAll(keyedWorkStreamThrottleTimes)
+        .build();
+  }
+
+  /** Starts {@link GetWorkerMetadataStream}. */
+  @SuppressWarnings({
+    "FutureReturnValueIgnored", // ignoring Future returned from Executor.submit()
+    "nullness" // Uninitialized value of getWorkerMetadataStream is null.
+  })
+  private void startGetWorkerMetadataStream() {
+    // We only want to set and start this value once.
+    if (getWorkerMetadataStream == null) {
+      synchronized (this) {
+        if (getWorkerMetadataStream == null) {
+          getWorkerMetadataStream =
+              streamingEngineStreamFactory.createGetWorkerMetadataStream(
+                  dispatcherClient.getDispatcherStub(),
+                  getWorkerMetadataThrottleTimer,
+                  endpoints ->
+                      consumeWorkerMetadataExecutor.submit(
+                          () -> consumeWindmillWorkerEndpoints(endpoints)));
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startBudgetRefreshThread() {
+    // Only 1 task will run at a time from the ScheduledExecutor documentation: If any
+    // execution of this task takes longer than its period, then subsequent executions may start
+    // late, but will not concurrently execute.
+    budgetRefreshExecutor.scheduleAtFixedRate(
+        this::refreshBudget,
+        // No need to initially wait since refreshBudget will wait before redistributing the budget.
+        0,
+        // Do not wait between successive executions. This should be constantly running.
+        0,
+        TimeUnit.MILLISECONDS);
+  }
+
+  private void refreshBudget() {
+    if (isBudgetRefreshPaused.get() || !started.get()) {
+      return;
+    }
+
+    // Phaser.awaitAdvanceInterruptibly() returns a negative value if the phaser is
+    // terminated.
+    if (waitForBudgetRefreshTrigger() < 0) {
+      return;
+    }
+
+    redistributeBudget();
+  }
+
+  private int waitForBudgetRefreshTrigger() {
+    try {
+      return budgetRefreshTrigger.awaitAdvanceInterruptibly(
+          0, SCHEDULED_BUDGET_REFRESH_MILLIS, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException | TimeoutException e) {
+      // If there is an error, we will proceed with refreshing the budget.
+      LOG.error("Error occurred waiting for budget refresh.", e);
```

Comment:
timeout should not be considered an error, we don't necessarily expect to be triggered more often than timeout.

Reply:
done